lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -803,17 +806,13 @@ namespace {
             // Copy the params and replace the parameter at each index with the
             // difference of the two spaces.  This unpacks one constructor head
             // into each parameter.
-            SmallVector<Space, 4> copyParams(this->getSpaces().begin(),
-                                             this->getSpaces().end());
             copyParams[idx] = s1.minus(s2, TC, DC);
-            Space CS = Space::forConstructor(this->getType(), this->getHead(),
-                                             this->canDowngradeToWarning(),
-                                             copyParams);
-            constrSpaces.push_back(CS);
           }
 
           if (foundBad) {
-            return Space::forDisjunct(constrSpaces);
+            return Space::forConstructor(this->getType(), this->getHead(),
```

Comment:
I don't know this code  well, so was surprised that it would be OK to return the constructor space instead of the disjunct space.

Reply:
That's right, instead of expanding `(A, B)` where `A { a, b }` and `B { c, d }` into series of embedded disjuncts (e.g. `((.a, .b), _: B) | (_: A, (.c, .d))` in process of subtraction, my change constructs a single disjunct of exhaustive set of constructors per tuple (e.g. `(.a, .c) | (.a, .d) | (.b, .c) | (.b, .d)`) as in the example from the description which helps to reduce covered space in subtraction because after each case is subtracted we'd end up with a bunch of empty spaces which is much easier to simplify comparing to current behavior. I'm going to add comments and prettify code some more when I'm convinced that everything works as expected and I didn't break some corner case by accident...

Reply:
This doesn't seem like a correct transformation to me. `(A, B) - (.a, .c)` should produce `(.a, .d) | (.b, .c) | (.b, .d)`, but the subtraction Robert wrote out would produce only `(.b, .d)`.

Reply:
FWIW it took me a long time to convince myself that the original code was correct and that the transformation you made was *not* correct when I went through this the first time, and that probably means there should be a (better?) comment.

Reply:
I agree that there should be a better comment but I don't understand why the transformation is _not_ correct, could you please elaborate? Here is my understanding - `(A, B)` is a type space, not a constructor space, it now get decomposed into a cartesian product of it's constructors - disjunction of `const(.a, .c) | const(.a, .d) | const(.b, .c) | const(.b, .d)` (where before it would be a constructor of two type sub-spaces `const(type(A), type(B))`) and when we subtract single constructor `const(.a, .c)` we'd always get a reduced constructor space and not a disjunction like before.

Reply:
That's what the `decomposeTuple` code is doing, but it's not what this change to `minus` is doing. If I actually *do* have `const(type(A), type(B))` and subtract `const(.a, .c)`, it is not possible for me to get a single constructor space out. That's just incorrect.

Reply:
But it is not possible now to get `const(type(A), type(A))` because tuples are expanded into cartesian product by decompose.

Reply:
You can still get it from an enum: `case foo(A, B)`.

Reply:
Also, it's not okay to implement the wrong algorithm without asserting that the inputs that would produce the wrong output don't show up. I'm not sure what those assertions would be, so you shouldn't do it.

Reply:
Is this really called for? As you can see in my original comment I'm trying to get some initial feedback for this patch which I don't yet consider complete. Regarding `case foo(A, B)` I think this the example:

```swift
enum A { case a }
enum B { case b, c }
enum C { case foo(A, B) }
func foo(_ c: C) {
  switch (c) {
    case .foo(_, .b):
      break
  }
}
```

I can see now that when decompose for enums uses the same mechanism as before (quiet literally identical) where it decomposes types into one space each, which can be changed to use `decomposeTuple` instead, also `projectPattern` doesn't have to decompose `TuplePattern` the way it does, it can do the same cartesian product, that would mean we'd get `const#foo(type(A, B))` decomposed into `const#foo(const(.a, .b), const(.a, .c))` and case pattern would be `const#foo((type(A), .b))` which is going to get expanded into `const#foo(const(.a, .b))`.

Also if we consider original algorithm of subtraction for constructor spaces using the same example after `foo(_, .b)` happened, it would result in `const#foo((disjunct(<empty>, ...), type(B)), (type(A), disjunct(<empty>, .c, ...)))` which doesn't seem to be better than `const#foo(const(<empty>, <empty>), const(<empty>, .c))` because it's much harder to simplify.

Reply:
I've talked to @jrose-apple offline and I think we've cleared up why it's only correct to create new constructors with all of the spaces of original (which effectively makes a single constructor a disjunct if one of the parameters is further divisible), one of the reasons being that constructor with empty parameter space doesn't really make sense and is effectively empty itself which means that for recursive enums we can't really construct cartesian product eagerly, unlike simpler cases were parameters are in their final indivisible form.